File: src/main/java/net/reflact/server/ReflactServer.java
package net.reflact.server;

import net.minestom.server.MinecraftServer;
import net.minestom.server.coordinate.Pos;
import net.minestom.server.entity.Player;
import net.minestom.server.event.GlobalEventHandler;
import net.minestom.server.event.player.AsyncPlayerConfigurationEvent;
import net.minestom.server.instance.anvil.AnvilLoader;
import net.minestom.server.instance.InstanceContainer;
import net.minestom.server.instance.InstanceManager;
import net.minestom.server.instance.LightingChunk;
import net.minestom.server.instance.block.Block;
import net.minestom.server.Auth;
import net.reflact.engine.ReflactEngine;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ReflactServer {
  private static final Logger LOGGER = LoggerFactory.getLogger(ReflactServer.class);

  public static void main(String[] args) {
    // Initialize the server with Online Mode
    MinecraftServer minecraftServer = MinecraftServer.init(new Auth.Online());

    // Initialize our engine
    ReflactEngine.init();

    InstanceManager instanceManager = MinecraftServer.getInstanceManager();
    // Use AnvilLoader for persistence. 'world' is the folder name.
    InstanceContainer instanceContainer = instanceManager.createInstanceContainer(new AnvilLoader("world"));
    instanceContainer.setGenerator(unit -> unit.modifier().fillHeight(0, 40, Block.GRASS_BLOCK));
    instanceContainer.setChunkSupplier(LightingChunk::new);

    GlobalEventHandler globalEventHandler = MinecraftServer.getGlobalEventHandler();
    globalEventHandler.addListener(AsyncPlayerConfigurationEvent.class, event -> {
      final Player player = event.getPlayer();
      event.setSpawningInstance(instanceContainer);
      player.setRespawnPoint(new Pos(0, 42, 0));
    });

    // Shutdown hook to save chunks
    MinecraftServer.getSchedulerManager().buildShutdownTask(() -> {
      LOGGER.info("Saving world...");
      instanceContainer.saveChunksToStorage();
    });

    // Initialize Server Console
    if (!java.awt.GraphicsEnvironment.isHeadless()) {
      new net.reflact.server.gui.ServerConsole().start();
    } else {
      LOGGER.info("Headless environment detected. GUI Console disabled.");
    }

    // Console handler (stdin fallback)
    new Thread(() -> {
      try (java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.InputStreamReader(System.in))) {
        String line;
        while ((line = reader.readLine()) != null) {
          if (line.trim().isEmpty())
            continue;
          MinecraftServer.getCommandManager().execute(MinecraftServer.getCommandManager().getConsoleSender(), line);
        }
      } catch (java.io.IOException e) {
        e.printStackTrace();
      }
    }, "Console-Thread").start();

    LOGGER.info("Starting server on port 25565 (Online Mode Enabled)...");
    minecraftServer.start("0.0.0.0", 25565);
  }
}

File: bin/main/net/reflact/server/ReflactServer.kt
package net.reflact.server

import net.minestom.server.MinecraftServer
import net.minestom.server.Auth
import net.minestom.server.coordinate.Pos
import net.minestom.server.event.player.AsyncPlayerConfigurationEvent
import net.minestom.server.instance.anvil.AnvilLoader
import net.minestom.server.instance.LightingChunk
import net.minestom.server.instance.block.Block
// import net.minestom.server.extras.MojangAuth
import net.reflact.engine.ReflactEngine
import org.slf4j.LoggerFactory
import java.io.BufferedReader
import java.io.InputStreamReader
import java.awt.GraphicsEnvironment
import net.reflact.server.gui.ServerConsole
import kotlin.concurrent.thread

object ReflactServer {
    private val LOGGER = LoggerFactory.getLogger(ReflactServer::class.java)

    @JvmStatic
    fun main(args: Array<String>) {
        // Initialize the server
        val minecraftServer = MinecraftServer.init(Auth.Online())

        // Initialize our engine
        ReflactEngine.init()

        val instanceManager = MinecraftServer.getInstanceManager()
        // Use AnvilLoader for persistence. 'world' is the folder name.
        val instanceContainer = instanceManager.createInstanceContainer(AnvilLoader("world"))
        instanceContainer.setGenerator { unit -> unit.modifier().fillHeight(0, 40, Block.GRASS_BLOCK) }
        instanceContainer.setChunkSupplier(::LightingChunk)

        val globalEventHandler = MinecraftServer.getGlobalEventHandler()
        globalEventHandler.addListener(AsyncPlayerConfigurationEvent::class.java) { event ->
            val player = event.player
            event.spawningInstance = instanceContainer
            player.respawnPoint = Pos(0.0, 42.0, 0.0)

            // Register channels so client knows it can send packets
            ReflactEngine.getNetworkManager().onPlayerConfiguration(player)
        }

        // Shutdown hook to save chunks
        MinecraftServer.getSchedulerManager().buildShutdownTask {
            LOGGER.info("Saving world...")
            instanceContainer.saveChunksToStorage().join()
        }

        // Spawn NPCs
        // NpcManager.spawnAll logic needs to be implemented or accessed correctly if added
        ReflactEngine.getNpcManager().spawnNpcs(instanceContainer)

        // Initialize Server Console
        if (!GraphicsEnvironment.isHeadless()) {
            ServerConsole().start()
        } else {
            LOGGER.info("Headless environment detected. GUI Console disabled.")
        }

        // Console handler (stdin fallback)
        thread(name = "Console-Thread") {
            try {
                BufferedReader(InputStreamReader(System.`in`)).use { reader ->
                    var line: String?
                    while (reader.readLine().also { line = it } != null) {
                        if (line!!.trim().isEmpty()) continue
                        MinecraftServer.getCommandManager().execute(MinecraftServer.getCommandManager().consoleSender, line!!)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        LOGGER.info("Starting server on port 25565 (Online Mode Enabled)...")
        minecraftServer.start("0.0.0.0", 25565)
    }
}
